epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2023 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2.io.epoll
+package unsafe
+
+import cats.effect.{IO, Resource, FileDescriptorPoller, FileDescriptorPollHandle}
+import cats.effect.std.Mutex
+import cats.syntax.all._
+import cats.effect.unsafe.PollingSystem
+
+import java.io.IOException
+
+import jnr.ffi.Memory;
+import jnr.constants.platform.Errno;
+
+import scala.collection.mutable.HashMap
+import scala.annotation.tailrec
+
+object EpollSystem extends PollingSystem {
+
+  import epoll._
+  import libc.jnr._
+
+  private[this] final val MaxEvents = 64
+
+  type Api = FileDescriptorPoller
+
+  def close(): Unit = ()
+
+  def makeApi(register: (Poller => Unit) => Unit): Api =
+    new FileDescriptorPollerImpl(register)
+
+  def makePoller(): Poller = {
+    val fd = epoll_create1(0)
+    if (fd == -1)
+      throw new IOException(strerror(errno()))
+    new Poller(fd)
+  }
+
+  def closePoller(poller: Poller): Unit = poller.close()
+
+  def poll(poller: Poller, nanos: Long, reportFailure: Throwable => Unit): Boolean =
+    poller.poll(nanos)
+
+  def needsPoll(poller: Poller): Boolean = poller.needsPoll()
+
+  def interrupt(targetThread: Thread, targetPoller: Poller): Unit = ()
+
+  private final class FileDescriptorPollerImpl private[EpollSystem] (
+      register: (Poller => Unit) => Unit
+  ) extends FileDescriptorPoller {
+    def registerFileDescriptor(
+        fd: Int,
+        reads: Boolean,
+        writes: Boolean
+    ): Resource[IO, FileDescriptorPollHandle] =
+      Resource {
+        (Mutex[IO], Mutex[IO]).flatMapN { (readMutex, writeMutex) =>
+          IO.async_[(PollHandle, IO[Unit])] { cb =>
+            register { epoll =>
+              val handle = new PollHandle(readMutex, writeMutex)
+              epoll.register(fd, reads, writes, handle, cb)
+            }
+          }
+        }
+      }
+  }
+
+  private final class PollHandle(
+      readMutex: Mutex[IO],
+      writeMutex: Mutex[IO]
+  ) extends FileDescriptorPollHandle {
+    private[this] var readReadyCounter = 0
+    private[this] var readCallback: Either[Throwable, Int] => Unit = null
+
+    private[this] var writeReadyCounter = 0
+    private[this] var writeCallback: Either[Throwable, Int] => Unit = null
```

Comment:
Some/all of these need to be `@volatile` in multi-threading. I need to think about it, for now we can mark them all `@volatile`.